hloc/match_features.py

```diff
@@ -90,74 +95,132 @@ def main(conf: Dict,
                      feature_paths_refs: Path,
-                     overwrite: bool = False) -> Path:
+                     overwrite: bool = False,
+                     
+                     device=None,
+                     num_workers=1,
+                     batch_size=1) -> Path:
+
     logger.info('Matching local features with configuration:'
                 f'\n{pprint.pformat(conf)}')
 
-    if not feature_path_q.exists():
-        raise FileNotFoundError(f'Query feature file {feature_path_q}.')
-    for path in feature_paths_refs:
-        if not path.exists():
-            raise FileNotFoundError(f'Reference feature file {path}.')
-    name2ref = {n: i for i, p in enumerate(feature_paths_refs)
-                for n in list_h5_names(p)}
-
     assert pairs_path.exists(), pairs_path
     pairs = parse_retrieval(pairs_path)
     pairs = [(q, r) for q, rs in pairs.items() for r in rs]
 
-    device = 'cuda' if torch.cuda.is_available() else 'cpu'
+    if device is None:
+      device = 'cuda' if torch.cuda.is_available() else 'cpu'
+
     Model = dynamic_load(matchers, conf['model']['name'])
     model = Model(conf['model']).eval().to(device)
 
     match_path.parent.mkdir(exist_ok=True, parents=True)
     skip_pairs = set(list_h5_names(match_path)
                      if match_path.exists() and not overwrite else ())
 
-    for (name0, name1) in tqdm(pairs, smoothing=.1):
-        pair = names_to_pair(name0, name1)
-        # Avoid to recompute duplicates to save time
-        if pair in skip_pairs or names_to_pair(name0, name1) in skip_pairs:
-            continue
-
-        data = {}
-        with h5py.File(str(feature_path_q), 'r') as fd:
-            grp = fd[name0]
-            for k, v in grp.items():
-                data[k+'0'] = torch.from_numpy(v.__array__()).float().to(device)
-            # some matchers might expect an image but only use its size
-            data['image0'] = torch.empty((1,)+tuple(grp['image_size'])[::-1])
-        with h5py.File(str(feature_paths_refs[name2ref[name1]]), 'r') as fd:
-            grp = fd[name1]
-            for k, v in grp.items():
-                data[k+'1'] = torch.from_numpy(v.__array__()).float().to(device)
-            data['image1'] = torch.empty((1,)+tuple(grp['image_size'])[::-1])
-        data = {k: v[None] for k, v in data.items()}
-
-        pred = model(data)
-        with h5py.File(str(match_path), 'a') as fd:
-            if pair in fd:
-                del fd[pair]
-            grp = fd.create_group(pair)
-            matches = pred['matches0'][0].cpu().short().numpy()
-            grp.create_dataset('matches0', data=matches)
-
-            if 'matching_scores0' in pred:
-                scores = pred['matching_scores0'][0].cpu().half().numpy()
-                grp.create_dataset('matching_scores0', data=scores)
-
-        skip_pairs.add(pair)
+    pairs = [pair for pair in set(pairs) 
+      if names_to_pair(*pair) not in skip_pairs]
```

Comment:
Here we need to also
1. check for existence in `skip_pairs` for both directions 0->1 and 1->0
2. Remove duplicated equivalent pairs

Reply:
Ah ha, thanks.

Reply:
I actually added this in PR https://github.com/cvg/Hierarchical-Localization/pull/159, I will merge it in the next days